pkg/registry/etcd/nse_server.go

```diff
@@ -19,31 +19,116 @@
 	"context"
 	"io"
 	"sync"
 	"time"
 
+	"github.com/edwarnicke/serialize"
 	"github.com/golang/protobuf/ptypes/empty"
-	"github.com/networkservicemesh/sdk/pkg/tools/log"
+
 	"github.com/pkg/errors"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/watch"
 
 	"github.com/networkservicemesh/api/pkg/api/registry"
 	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
 	"github.com/networkservicemesh/sdk/pkg/tools/matchutils"
 
 	v1 "github.com/networkservicemesh/sdk-k8s/pkg/tools/k8s/apis/networkservicemesh.io/v1"
 	"github.com/networkservicemesh/sdk-k8s/pkg/tools/k8s/client/clientset/versioned"
 )
 
 type etcdNSERegistryServer struct {
-	chainContext context.Context
-	client       versioned.Interface
-	versions     sync.Map
-	ns           string
+	chainContext   context.Context
+	deleteExecutor serialize.Executor
+	client         versioned.Interface
+	versions       sync.Map
+	ns             string
+
+	subscribers      *list.List
+	subscribersMutex sync.Mutex
+
+	updateChannelSize int
+}
+
+// NewNetworkServiceEndpointRegistryServer creates new registry.NetworkServiceRegistryServer that is using etcd to store network services.
+func NewNetworkServiceEndpointRegistryServer(chainContext context.Context, ns string, client versioned.Interface) registry.NetworkServiceEndpointRegistryServer {
+	ret := &etcdNSERegistryServer{
+		chainContext:      chainContext,
+		client:            client,
+		ns:                ns,
+		subscribers:       list.New(),
+		updateChannelSize: 64,
+	}
+
+	go ret.watchRemoteStorage()
+
+	return ret
+}
+
+func (n *etcdNSERegistryServer) watchRemoteStorage() {
+	const minSleepTime = time.Millisecond * 50
+	const maxSleepTime = minSleepTime * 5
+
+	sleepTime := minSleepTime
+	logger := log.FromContext(n.chainContext).WithField("etcdNSERegistryServer", "watchRemoteStorage")
+
+	for n.chainContext.Err() == nil {
+		timeoutSeconds := int64(time.Minute / time.Second)
+		watcher, err := n.client.NetworkservicemeshV1().NetworkServiceEndpoints("").Watch(n.chainContext, metav1.ListOptions{
+			TimeoutSeconds: &timeoutSeconds,
+		})
+		if err != nil {
+			time.Sleep(sleepTime)
+			sleepTime += minSleepTime
+			sleepTime = min(sleepTime, maxSleepTime)
+			continue
+		}
+		sleepTime = max(sleepTime, minSleepTime)
+
+		isWatcherFine := true
+		for isWatcherFine {
+			select {
+			case <-n.chainContext.Done():
+				return
+			case event, ok := <-watcher.ResultChan():
+				if !ok {
+					isWatcherFine = false
+					break
+				}
+				deleted := event.Type == watch.Deleted
+				model, ok := event.Object.(*v1.NetworkServiceEndpoint)
+				if !ok {
+					logger.Errorf("event: %v", event)
+					continue
+				}
+				item := (*registry.NetworkServiceEndpoint)(&model.Spec)
+				if item.Name == "" {
+					item.Name = model.GetName()
+				}
+				if v, ok := n.versions.Load(item.Name); ok && v == model.ResourceVersion {
```

Comment:
I don't understand the purpose of this check.
In case the NSE is created/updated via a different registry then it's pointless. While in case the same registry takes care of the registration that also serves "find subscribers" of an NSE, then why risk race between watchRemoteStorage thread and register (jeopardizing sending out a notification to old subscribers about a new NSE for example)?

Reply:
Agreed, seems like it's not needed when we change the logic with the watcher, so just removed it. Thanks!